cranelift/codegen/src/isa/x64/lower.isle

```diff
@@ -964,3 +964,60 @@
 ;; all-one value
 (rule (lower (has_type ty @ (multi_lane _bits _lanes) (bnot x)))
       (value_reg (sse_xor ty (put_in_reg x) (RegMem.Reg (vector_all_ones ty)))))
+
+;;;; Rules for `insertlane` ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(rule (lower (insertlane vec @ (value_type ty) val (u8_from_uimm8 idx)))
```

Comment:
It occurs to me looking at this that I should have originally put a `debug_assert` in the lowering code to check for correct indexes; I don't really think this PR needs to change but I'm curious--if we did want to check that the index was in range for the type, how would we do so in ISLE?

Reply:
I don't think that a `debug_asert!` would directly translate but what could be done is to have a different extractor here instead of `u8_from_uimm8` but rather something like `u{1,2,3,4}_from_uimm8` where the rule wouldn't get matched if the index was out-of-bounds, and then later we'd have a codegen error of "ISLE didn't match this" and manual inspection would show that the lane index was out-of-bounds. @fitzgen may have other ideas though.

Reply:
Ok, cool; thanks, just curious.

Reply:
The right way would probably be to assert in Rust code called by the rule; either in one of the extractors in the pattern (make a specific `LaneIndex` type then a `(lane_index_from_uimm8 idx)` extractor pattern that returns it?) or in the constructor of whatever instruction takes it.

We could also have a generic `(assert ...)` constructor that we could invoke in the right-hand side but that seems like a bit of an antipattern; better to just encode the invariant in the types :-)

Reply:
Seconding what @cfallin said :)